### PR TITLE
Podcast Player: implement colors in server-side

### LINKED ISSUE #15194: Podcast Player: Apply colors on the server-side
Apply colors in the server-side by adding CSS classes and inline style to the HTML markup.

### PR DESCRIPTION
**Pull request by retrofox**
This PR implements colors in the server-side rendering by adding CSS classes and inline styles, depending on the settings.

Fixes https://github.com/Automattic/jetpack/issues/15194

#### Changes proposed in this Pull Request:

* Build CSS classes and inline styles based on the block attributes.
* Apply these values straight to the markup built in the server-side

#### Testing instructions:

1) Edit/create a Podcast Player form the editor-canvas side.
2) Appy colors using the inspector controls (sidebar). Try to pick up colors from the palette as well as defining custom colors:

Screen Shot 2020-03-30 at 11 18 37 AM

3) Save the post

In the front-end, let's check how it renders disabling javascript. For this:

a) Open Chrome DevTools.
b) Press Control+Shift+P or Command+Shift+P (Mac) to open the Command Menu.
c) Start typing javascript, select Disable JavaScript, and then press Enter to run the command. JavaScript is now disabled.
image

4) Hard Refresh (always in the front-end). The player shouldn't work since JS has been disabled.
5) Check CSS classes and inline are applied, for instance:

image

image

#### Proposed changelog entry for your changes:

* Apply colors in markup server-side rendering

### CODE REVIEW

File: extensions/blocks/podcast-player/podcast-player.php [hunk 1]
```diff
@@ -115,28 +115,81 @@ function render_player( $player_data, $attributes ) {
...
+	$background_classes_name = '';
+	$background_inline_style = '';
+	if ( isset( $background_color_class ) || isset( $custom_background_color ) ) {
+		$background_classes_name .= ' has-background';
+		if ( isset( $background_color_class ) ) {
+			$background_classes_name .= $background_color_class;
+		} elseif ( isset( $custom_background_color ) ) {
+			$background_inline_style .= "background-color: $custom_background_color;";
+		}
+	}
```
Comment (obenland) [inline extensions/blocks/podcast-player/podcast-player.php]: It would probably be easier to read if all of the above was in a separate function

File: extensions/blocks/podcast-player/podcast-player.php [hunk 2]
```diff
@@ -161,3 +214,19 @@ class="jetpack-podcast-player__episode-link"
...
+ * @param string $color_slug         Slug of the color.
+ *
+ * @return string String with the class corresponding to the color in the provided context.
+ */
+function get_color_class_name( $color_context_name, $color_slug ) {
+	if ( ! isset( $color_context_name ) || ! isset( $color_slug ) ) {
+		return null;
+	}
+
+	return "has-${color_slug}-${color_context_name}";
```
[the remark this answers is not in the document]
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: Actually, it works in both ways. Anyway going to update them if you consider it's better.

File: extensions/blocks/podcast-player/podcast-player.php [hunk 3]
```diff
@@ -161,3 +214,19 @@ class="jetpack-podcast-player__episode-link"
...
+ * Returns a CSS class based on the context a color is being used and its slug.
+ *
+ * @param string $color_context_name Context/place where color is being used e.g: background, text etc...
+ * @param string $color_slug         Slug of the color.
+ *
+ * @return string String with the class corresponding to the color in the provided context.
+ */
+function get_color_class_name( $color_context_name, $color_slug ) {
+	if ( ! isset( $color_context_name ) || ! isset( $color_slug ) ) {
+		return null;
```
[the remark this answers is not in the document]
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: In that case, I'll have to change `isset` instead of `empty`. Changing...

File: extensions/blocks/podcast-player/podcast-player.php [hunk 4]
```diff
@@ -115,28 +115,81 @@ function render_player( $player_data, $attributes ) {
...
-				<a
-					class="jetpack-podcast-player__episode-link"
-					href="<?php echo esc_url( $attachment['link'] ); ?>"
-					role="button"
-					aria-pressed="false"
+		<section
+			<?php
+			echo ! empty( $podcast_player_classes_name )
+				? ' class="' . esc_attr( $podcast_player_classes_name ) . '"'
+				: '';
```
[the remark this answers is not in the document]
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: I don't like rendering empty attributes, to be honest. That's why I did it in this not-fancy way.

File: extensions/blocks/podcast-player/podcast-player.php (hunk 1 above)
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: Not sure. We've worked previously in the Navigation menu and honestly, I'm don't think that splitting the code out in helper functions will improve the readability.

File: extensions/blocks/podcast-player/podcast-player.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (obenland) [reply extensions/blocks/podcast-player/podcast-player.php]: Huh, I did not realize that. I wonder if that is a bug or feature :) 
Feel free not to change it if it works

File: extensions/blocks/podcast-player/podcast-player.php [hunk 5]
```diff
@@ -115,28 +115,81 @@ function render_player( $player_data, $attributes ) {
...
+		if ( isset( $secondary_color_class ) ) {
+			$secondary_classes_name .= $secondary_color_class;
+		} elseif ( isset( $custom_secondary_color ) ) {
+			$secondary_inline_style .= "color: $custom_secondary_color;";
+		}
+	}
+
+	$background_classes_name = '';
+	$background_inline_style = '';
+	if ( isset( $background_color_class ) || isset( $custom_background_color ) ) {
```
[the remark this answers is not in the document]
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: sorry, what do you mean

File: extensions/blocks/podcast-player/podcast-player.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: hahaha, to me it's a feature!

File: extensions/blocks/podcast-player/podcast-player.php [hunk 6]
```diff
@@ -115,28 +115,70 @@ function render_player( $player_data, $attributes ) {
...
 	<div class="<?php echo esc_attr( $block_classname ); ?>" id="<?php echo esc_attr( $instance_id ); ?>">
-		<ol class="jetpack-podcast-player__episodes">
-			<?php foreach ( $player_data['tracks'] as $attachment ) : ?>
-			<li class="jetpack-podcast-player__episode">
-				<a
-					class="jetpack-podcast-player__episode-link"
-					href="<?php echo esc_url( $attachment['link'] ); ?>"
-					role="button"
-					aria-pressed="false"
+		<section
```
Comment (jeherve) [inline extensions/blocks/podcast-player/podcast-player.php]: Do we need that additional level with `section`, or could we roll this in the parent `div` instead?

**Review by jeherve — CHANGES_REQUESTED** [review]
Maybe we could take an approach similar to what we did for Calendly, to simplify things a bit?
https://github.com/Automattic/jetpack/blob/70dab921662987d1dd1aa7e0f4f302dc692989d2/extensions/blocks/calendly/calendly.php#L178-L186

File: extensions/blocks/podcast-player/podcast-player.php (hunk 6 above)
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: What I' trying to do here is rendering the HTML code as similar as possible [the client does](https://github.com/Automattic/jetpack/blob/master/extensions/blocks/podcast-player/components/podcast-player.js#L176).

File: extensions/blocks/podcast-player/podcast-player.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (obenland) [reply extensions/blocks/podcast-player/podcast-player.php]: `if ( $background_color_class || $custom_background_color ) {`

**Comment by retrofox** [comment]
> Maybe we could take an approach similar to what we did for Calendly, to simplify things a bit?
> https://github.com/Automattic/jetpack/blob/70dab921662987d1dd1aa7e0f4f302dc692989d2/extensions/blocks/calendly/calendly.php#L178-L186

Hi Jeremy, not sure how we can apply that similar approach here. For instance, we aren't defining default values at least for now. Going to add some helper functions in order to try to make the code clearer.

File: extensions/blocks/podcast-player/podcast-player.php (hunk 6 above)
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: so the answer is `yes` :-D

File: extensions/blocks/podcast-player/podcast-player.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (obenland) [reply extensions/blocks/podcast-player/podcast-player.php]: Sorry, I seem to have posted that answer to the wrong conversation. I thought I was adding to the variable-in-string conversation. 

I still think we should change this to return an empty string. You don't necessarily have to add an `empty()` check since `''` evaluates as falsy.

File: extensions/blocks/podcast-player/podcast-player.php [hunk 7]
```diff
@@ -115,28 +115,42 @@ function render_player( $player_data, $attributes ) {
 		$player_data
 	);
 
+	$secondary_colors  = get_colors( 'secondary', $attributes, 'color' );
+	$background_colors = get_colors( 'background', $attributes, 'background-color' );
```
Comment (obenland) [inline extensions/blocks/podcast-player/podcast-player.php]: Let's pass `secondaryColor` and `backgroundColor` as the first parameter to make it just a little less closely coupled.

File: extensions/blocks/podcast-player/podcast-player.php [hunk 8]
```diff
@@ -115,28 +115,42 @@ function render_player( $player_data, $attributes ) {
...
+
+	$player_classes_name = trim( "{$secondary_colors['class']} {$background_colors['class']}" );
+	$player_inline_style = trim( "{$secondary_colors['style']} ${background_colors['style']}" );
```
Comment (obenland) [inline extensions/blocks/podcast-player/podcast-player.php]: Do we actually need to trim those attributes here? Looks to me like `get_colors()` doesn't add unnecessary whitespace

File: extensions/blocks/podcast-player/podcast-player.php [hunk 9]
```diff
@@ -161,3 +175,43 @@ class="jetpack-podcast-player__episode-link"
 
 	return ob_get_clean();
 }
+
+/**
+ * Given the color name, bock attributes and the CSS property,
```
Comment (obenland) [inline extensions/blocks/podcast-player/podcast-player.php]: s/bock/block

File: extensions/blocks/podcast-player/podcast-player.php [hunk 10]
```diff
@@ -161,3 +175,43 @@ class="jetpack-podcast-player__episode-link"
...
+ * the function will return an array with the `class` and `style`
+ * HTML attributes to be used straight in the markup.
+ *
+ * @example
+ * $color = get_colors( 'secondary', $attributes, 'border-color'
+ *  => array( 'class' => 'has-secondary', 'style' => 'border-color: #333' )
+ *
+ * @param string $name     Color attribute name, for instance `primary`, `secondary`, ...
+ * @param array  $attrs     Block attributes.
```
Comment (obenland) [inline extensions/blocks/podcast-player/podcast-player.php]: Looks like the attribute description is off by one space

File: extensions/blocks/podcast-player/podcast-player.php [hunk 11]
```diff
@@ -161,3 +175,43 @@ class="jetpack-podcast-player__episode-link"
...
+	$color        = isset( $attrs[ $attr_color ] ) ? $attrs[ $attr_color ] : null;
+	$custom_color = isset( $attrs[ $attr_custom ] ) ? $attrs[ $attr_custom ] : null;
+
+	$colors = array(
+		'class' => '',
+		'style' => '',
+	);
+
+	if ( $color || $custom_color ) {
+		$colors['class'] .= " has-{$name}";
```
Comment (obenland) [inline extensions/blocks/podcast-player/podcast-player.php]: Do we need that leading whitespace here?

File: extensions/blocks/podcast-player/podcast-player.php [hunk 12]
```diff
@@ -161,3 +175,43 @@ class="jetpack-podcast-player__episode-link"
...
+
+		if ( $color ) {
+			$colors['class'] .= ' ' . "has-{$property}-{$color}";
```
Comment (obenland) [inline extensions/blocks/podcast-player/podcast-player.php]: This could probably be simplified to `" has-{$property}-{$color}"` (with leading whitespace).

File: extensions/blocks/podcast-player/podcast-player.php [hunk 13]
```diff
@@ -161,3 +175,43 @@ class="jetpack-podcast-player__episode-link"
...
+		} elseif ( $custom_color ) {
+			$colors['style'] .= "{$property}: {$custom_color};";
+		}
+	}
+
+	$colors['class'] = trim( $colors['class'] );
```
Comment (obenland) [inline extensions/blocks/podcast-player/podcast-player.php]: Not sure we need to trim with above changes. If we do, this could work as well: `return array_map( 'trim', $color );`.

File: extensions/blocks/podcast-player/podcast-player.php (hunk 8 above)
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: if there are not color it will apply an empty string.
```html
<div class=" " style=" " />
```

File: extensions/blocks/podcast-player/podcast-player.php (hunk 11 above)
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: yes, we are contacting two string.

File: extensions/blocks/podcast-player/podcast-player.php (hunk 12 above)
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: yes....

File: extensions/blocks/podcast-player/podcast-player.php (hunk 11 above)
Comment (obenland) [reply extensions/blocks/podcast-player/podcast-player.php]: Connecting to an empty string though, no?

File: extensions/blocks/podcast-player/podcast-player.php (hunk 13 above)
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: > Not sure we need to trim with above changes

it's coating the CSS classes, so before to return these values I remove the spaces in extremes.

> If we do, this could work as well: return array_map( 'trim', $color );.
👍

File: extensions/blocks/podcast-player/podcast-player.php (hunk 8 above)
Comment (obenland) [reply extensions/blocks/podcast-player/podcast-player.php]: Oh, because `$secondary_colors['style']` or `$secondary_colors['class']` could be empty

File: extensions/blocks/podcast-player/podcast-player.php (hunk 7 above)
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: If we change that then this implementation will get more complex. The function uses the first argument to define `has-{$name}`:

```php
$color = get_colors( 'secondary', $attributes, 'border-color' )'

array(
    'class' => 'has-seconday' // <- this class is generated based on the color name
    'style' => 'border-color: #65ac45'
);
```

Comment (obenland) [reply extensions/blocks/podcast-player/podcast-player.php]: Ah, I didn't catch that. Sounds good 👍

File: extensions/blocks/podcast-player/podcast-player.php (hunk 8 above)
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: also both of them

File: extensions/blocks/podcast-player/podcast-player.php (hunk 11 above)
Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: this space together with `trim()` simplifies the process to handle these strings. We could use `array_push()` and at the end `implode()` it but I think it's almost the same.

Comment (retrofox) [reply extensions/blocks/podcast-player/podcast-player.php]: done!! sorry about this, thanks 🙏

**Review by obenland — APPROVED** [review]
Nice work

**Comment by retrofox** [comment]
> Nice work

🙇

**Review by jeherve — APPROVED** [review]

**Comment by retrofox** [comment]
r205104-wpcom
